### DISCUSSION TITLE
[maybe bug] ModuleFederation seems not compatible with ArrayPushCallbackChunkFormatPlugin, causing ScriptExternalLoadError

### DISCUSSION THREAD

**Discussion by tachish** [webpack/webpack#18308] [Q&A]
I find host project will throw `ScriptExternalLoadError` error while loading remote module sometimes, if remote project uses `splitChunkPlugin` or `runtime: 'single'` to optimize.

### Surface
This part will explain why `ScriptExternalLoadError` would be thrown.

In general, `remoteEntry.js` is an `IIFE` in remote project's bundle artifacts, looks like:
```
var remoteApp; // module federation's global variable
(() => {
"use strict";
// webpack runtime implementation here.

// module federation implementation here.

remoteApp = __webpack_exports__; // assign to module federation's global variable
})();
```

In host project, will use `__webpack_require.l` to load remote module. And at `__webpack_require__.l`'s callback, if module federation's global variable is nil, will throw `ScriptExternalLoadError`.
```
__webpack_require__.l("http://localhost:3002/remoteEntry.js", (event) => {
        if(typeof remoteApp !== "undefined") return resolve();
        var errorType = event && (event.type === 'load' ? 'missing' : event.type);
        var realSrc = event && event.target && event.target.src;
        __webpack_error__.message = 'Loading script failed.\n(' + errorType + ': ' + realSrc + ')';
        __webpack_error__.name = 'ScriptExternalLoadError';
        __webpack_error__.type = errorType;
        __webpack_error__.request = realSrc;
        reject(__webpack_error__);
}, "remoteApp");
```

However, `remoteEntry.js` is arrayfied if using `splitChunkPlugin` or `runtime: 'single'`, looks like:
```
"use strict";
var remoteApp; // module federation's global variable
(self["webpackChunkXXX"] = self["webpackChunkXXX"] || []).push([
        [334],
        {
                503: ((__unused_webpack_module, exports, __webpack_require__) => {
                        // module federation's implementation here.
                })
        },

        // following won't be executed after loading by host, remoteApp won't be assigned
        __webpack_require__ => { // webpackRuntimeModules
                var __webpack_exec__ = (moduleId) => (__webpack_require__(__webpack_require__.s = moduleId))
                var __webpack_exports__ = (__webpack_exec__(503));
                remoteApp = __webpack_exports__;
        }
]);
```

In this case, module federation's global variable is always nil, because assign expression is not executed after loading `remoteEntry.js`. It makes `__webpack_require__.l`'s callback always throw `ScriptExternalLoadError`.

### Hope to do
If it is actually a bug, I have a preliminary idea to fix it. If it is possible to keep module federation's entry file always be `IIFE`, instead of arrayfied sometimes.

**Comment by alexander-akait** [comment]
You need to preload them too... Can you provide a small example of your problem using github and I will show you how to fix and why it happend

**Reply by tachish** [reply]
Thanks for reply! I will provide an example over the weekend, appreciate again!
